### PR TITLE
privacy: Type privacy lints fixes and cleanups

### PR DESCRIPTION
**Pull request by petrochenkov**
See individual commits.
Follow up to https://github.com/rust-lang/rust/pull/111801.

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @Bryanskiy

**Comment by petrochenkov** [comment]
Ping @eholk, it's been 2 weeks.
The feature gating part should land before the next release.

File: compiler/rustc_lint/src/builtin.rs [hunk 1]
```diff
@@ -665,9 +665,7 @@ declare_lint_pass!(MissingCopyImplementations => [MISSING_COPY_IMPLEMENTATIONS])
 
 impl<'tcx> LateLintPass<'tcx> for MissingCopyImplementations {
     fn check_item(&mut self, cx: &LateContext<'_>, item: &hir::Item<'_>) {
-        if !(cx.effective_visibilities.is_reachable(item.owner_id.def_id)
-            && cx.tcx.local_visibility(item.owner_id.def_id).is_public())
-        {
+        if !cx.effective_visibilities.is_reachable(item.owner_id.def_id) {
```
Comment (eholk) [inline compiler/rustc_lint/src/builtin.rs:668]: Is the local visibility check just redundant because effective visibility is always stricter?

**Review by eholk — APPROVED** [review]
Sorry about the slow review! Thanks for breaking this out into different commits, that made it much easier to review.

@bors r+

File: compiler/rustc_lint/src/builtin.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_lint/src/builtin.rs:668]: Yes, with 95a24c6ed489de0fa6e87ef99390a744c8d46ddc reachable to other crates implies `pub` again, so we revert various "adaptation" changes from https://github.com/rust-lang/rust/pull/111801 too.

**Pull request merged**